### PR TITLE
feat(avalanche): integration of avalanche

### LINKED ISSUE #4864: [Bounty] Integrate Avalanche into Mask
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [ ] Intermediate
  - [x] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  [x] 4 weeks

- Bounty Size
  [x] 5000 USD

### Part 1: Problem Description

**Background** Avalanche is a flexible smart contract platform for creating custom blockchain networks and decentralized applications (especially those focused on DeFi). Avalanche facilitates the creation of custom blockchains (including private blockchains) and decentralized applications (dApps). Its native,fixed-supply AVAX coin is the platform’s unit of account, used for paying transaction fees and for staking in Avalanche’s consensus process.

**Solution** As the first step of our cooperation, we, Mask Network, would like to integrate Avalanche as one of our default supported blockchain system in our internal Mask Wallet, so that further on-chain interactions would happen directly within Mask Network without leaving the platform. We believe this would further help us attract more users from different projects, aiming at different goals to maximize users experiences in the broad blockchain world.
If you are interested in this bounty, please make sure that you contact us first to see how our structure looks like and proceed to the development afterwards.

### Part 2: Resources

- https://www.avax.network/
- https://github.com/DimensionDev/Maskbook/tree/master/packages/maskbook/src/plugins/Wallet

### Part 3: Contact and Support

@yisiliu for ideas @Jack-Works for implementations Mask Network discord server: https://discord.gg/4Xccj7a75Q Avalanche discord server: https://discord.com/invite/RwXY7P6

### LINKED ISSUE #4874: [Demand] Introduce Avalanche Chain
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

+ DEX: https://app.pangolin.exchange/#/swap (An uniswap-v2 like DEX.)
+ CoinGecko ID: `"avalanche"`
+ CMC ID: `"5805"`
+ Debank ID: `"avalanche"`
+ OpenSea: Not Supported
+ Explorer: https://explorer.avax.network/
+ RPC: https://api.avax.network/ext/bc/C/rpc

## Notes

+ [AVAX](https://www.avax.network/)
+ [Avalanche  Explorer](https://explorer.avax.network/)

### PR DESCRIPTION
**Pull request by amritkumarj**
## Description

Closes #4864
Closes #4874

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

### CODE REVIEW

File: packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx [hunk 1]
```diff
@@ -227,9 +227,15 @@ export function ClaimAllDialog(props: ClaimAllDialogProps) {
...
-        )
+        [
+            ChainId.Mainnet,
+            ChainId.BSC,
+            ChainId.Matic,
+            ChainId.Arbitrum,
+            ChainId.xDai,
+            ChainId.Fantom,
+            ChainId.Avalanche,
+        ].includes(currentChainId)
```
Comment (septs) [inline packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx]: extract the array as constants.

e.g: `ALLOWED_CHAIN_ID_LIST`

File: packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx [hunk 2]
```diff
@@ -375,6 +381,7 @@ export function ClaimAllDialog(props: ClaimAllDialogProps) {
                                 ChainId.Arbitrum,
                                 ChainId.xDai,
                                 ChainId.Fantom,
+                                ChainId.Avalanche,
```
Comment (septs) [inline packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx]: https://github.com/DimensionDev/Maskbook/pull/5348/files#r776580361

use extracted constant replace the `chains` hardcoded array.

Comment (amritkumarj) [reply packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx]: okay

Comment (amritkumarj) [reply packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx]: done

File: packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx (hunk 1 above)
Comment (amritkumarj) [reply packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx]: done

**Comment by DhruvJain1122** [comment]
@septs please check for any other changes

File: packages/mask/src/plugins/Trader/pipes.ts [hunk 3]
```diff
@@ -142,6 +144,8 @@ export function resolveTradePairLink(tradeProvider: TradeProvider, address: stri
                     return `https://analytics-celo.sushi.com/pairs/${address}`
                 case NetworkType.Fantom:
                     return `https://analytics-ftm.sushi.com/pairs/${address}`
+                case NetworkType.Avalanche:
+                    return ``
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/pipes.ts]: Please use the correct link here.

File: packages/mask/src/plugins/Trader/constants/sushiswap.ts [hunk 4]
```diff
@@ -35,4 +35,5 @@ export const SUSHISWAP_BASE_AGAINST_TOKENS: ERC20AgainstToken = {
     [ChainId.xDai]: [WNATIVE, USDC, USDT, WBTC].map((x) => x[ChainId.xDai]),
     [ChainId.Celo]: [WNATIVE, CUSD, CEUR].map((x) => x[ChainId.Celo]),
     [ChainId.Fantom]: [WNATIVE, DAI, USDC, fUSDT, WBTC].map((x) => x[ChainId.Fantom]),
+    [ChainId.Avalanche]: [WNATIVE, DAI, USDC, fUSDT, WBTC].map((x) => x[ChainId.Avalanche]),
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/constants/sushiswap.ts]: Please use the correct tokens here. You might need to find them from the sushiswap codebase.

File: packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx [hunk 5]
```diff
@@ -220,18 +220,23 @@ export function ClaimAllDialog(props: ClaimAllDialogProps) {
     const DialogRef = useRef<HTMLDivElement>(null)
     const account = useAccount()
     const currentChainId = useChainId()
+    const ALLOWED_CHAIN_ID_LIST = [
```
Comment (nuanyang233) [inline packages/mask/src/plugins/ITO/SNSAdaptor/ClaimAllDialog.tsx]: Plz move it outside the function

File: packages/web3-constants/evm/coinmarketcap.json [hunk 6]
```diff
@@ -13,6 +13,7 @@
         "Arbitrum_Rinkeby": "",
         "xDai": "5601",
         "Celo": "5567",
-        "Fantom": "3513"
+        "Fantom": "3513",
+        "Avalanche": "43114"
```
Comment (guanbinrui) [inline packages/web3-constants/evm/coinmarketcap.json]: It isn't a chain id. Learn more: https://github.com/DimensionDev/Maskbook/issues/4874

File: packages/web3-constants/evm/debank.json [hunk 7]
```diff
@@ -13,6 +13,7 @@
         "Arbitrum_Rinkeby": "",
         "xDai": "xdai",
         "Celo": "celo",
-        "Fantom": "ftm"
+        "Fantom": "ftm",
+        "Avalanche": "avax"
```
Comment (guanbinrui) [inline packages/web3-constants/evm/debank.json]: Learn more: #4874

File: packages/web3-constants/evm/explorer.json [hunk 8]
```diff
@@ -13,6 +13,7 @@
         "Arbitrum_Rinkeby": "",
         "xDai": "",
         "Celo": "",
+        "Fantom": "https://api.ftmscan.com/api",
         "Fantom": "https://api.ftmscan.com/api"
```
Comment (guanbinrui) [inline packages/web3-constants/evm/explorer.json]: Please update the key name. And use the correct value.

File: packages/web3-constants/evm/explorer.json [hunk 9]
```diff
@@ -29,6 +30,7 @@
         "Arbitrum_Rinkeby": "",
         "xDai": "",
         "Celo": "",
+        "Fantom": "",
         "Fantom": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/explorer.json]: Please update the key name.

File: packages/web3-constants/evm/gitcoin.json [hunk 10]
```diff
@@ -29,6 +30,7 @@
         "Arbitrum_Rinkeby": "",
         "xDai": "",
         "Celo": "",
+        "Avalanche": "",
```
Comment (guanbinrui) [inline packages/web3-constants/evm/gitcoin.json:36]: Sort `Avalanche` after `Fantom`. 

```bash
cd packages/web3-constants
npm run start
```

File: packages/web3-constants/evm/gitcoin.json [hunk 11]
```diff
@@ -45,6 +47,7 @@
...
```
Comment (guanbinrui) [inline packages/web3-constants/evm/gitcoin.json:50]: Sort `Avalanche` after `Fantom`.

File: packages/web3-constants/evm/gitcoin.json [hunk 12]
```diff
@@ -61,6 +64,7 @@
         "Arbitrum_Rinkeby": 5,
         "xDai": 5,
         "Celo": 0,
+        "Avalanche": 0,
```
Comment (guanbinrui) [inline packages/web3-constants/evm/gitcoin.json:67]: Sort `Avalanche` after `Fantom`.

File: pnpm-lock.yaml [hunk 13]
```diff
@@ -10575,8 +10575,8 @@ packages:
     engines: {node: ^6 || ^7 || ^8 || ^9 || ^10 || ^11 || ^12 || >=13.7}
     hasBin: true
     dependencies:
-      caniuse-lite: 1.0.30001286
-      electron-to-chromium: 1.4.16
+      caniuse-lite: 1.0.30001292
```
Comment (guanbinrui) [inline pnpm-lock.yaml]: Revert changes to this file.

File: .pnpmfile.cjs [hunk 14]
```diff
@@ -14,7 +14,7 @@ const approvedList = new Map([
         'wyvern-js',
         ['git+https://github.com/ProjectOpenSea/wyvern-js.git#v3.2.1', 'github:ProjectOpenSea/wyvern-js#semver:^3.2.1'],
     ],
-    ['wyvern-schemas', 'git+https://github.com/ProjectOpenSea/wyvern-schemas.git#v0.11.1'],
+    ['wyvern-schemas', 'git+https://github.com/ProjectOpenSea/wyvern-schemas.git#v0.13.1'],
```
Comment (septs) [inline .pnpmfile.cjs]: CC @Jack-Works

File: packages/mask/package.json [hunk 15]
```diff
@@ -2,7 +2,9 @@
   "name": "@masknet/extension",
   "private": true,
   "license": "AGPL-3.0-or-later",
-  "scripts": {},
+  "scripts": {
```
Comment (septs) [inline packages/mask/package.json]: plz revert the changes

File: packages/mask/package.json [hunk 16]
```diff
@@ -146,4 +148,4 @@
   "peerDependencies": {
     "tss-react": "*"
   }
-}
+}
```
Comment (septs) [inline packages/mask/package.json]: plz revert the change

File: packages/mask/package.json [hunk 17]
```diff
@@ -91,7 +93,7 @@
     "jsx-jsonml-devtools-renderer": "^1.4.3",
     "millify": "^4.0.0",
     "next-tick": "^1.0.0",
-    "opensea-js": "^1.1.11",
+    "opensea-js": "^1.2.7",
```
Comment (septs) [inline packages/mask/package.json]: plz revert the change?

CC @guanbinrui

File: packages/mask/src/components/shared/ApplicationBoard.tsx [hunk 18]
```diff
@@ -289,6 +289,7 @@ export function ApplicationBoard({ secondEntries, secondEntryChainTabs }: MaskAp
                         ChainId.xDai,
                         ChainId.Celo,
                         ChainId.Fantom,
+                        ChainId.Avalanche,
```
Comment (septs) [inline packages/mask/src/components/shared/ApplicationBoard.tsx]: extract the chain id list as constant

File: packages/web3-shared/evm/pipes/index.ts [hunk 19]
```diff
@@ -79,6 +80,7 @@ export const resolveNetworkName = createLookupTableResolver<NetworkType, string>
         [NetworkType.xDai]: 'xDai',
         [NetworkType.Celo]: 'Celo',
         [NetworkType.Fantom]: 'Fantom',
+        [NetworkType.Avalanche]: 'avalanche',
```
Comment (guanbinrui) [inline packages/web3-shared/evm/pipes/index.ts]:
```suggestion
        [NetworkType.Avalanche]: 'Avalanche',
```

**Review by septs — CHANGES_REQUESTED** [review]

File: .pnpmfile.cjs (hunk 14 above)
Comment (Jack-Works) [reply .pnpmfile.cjs]: It's ok I guess.

File: packages/web3-constants/evm/explorer.json [hunk 20]
```diff
@@ -13,7 +13,8 @@
         "Arbitrum_Rinkeby": "",
         "xDai": "",
         "Celo": "",
-        "Fantom": "https://api.ftmscan.com/api"
+        "Fantom": "https://api.ftmscan.com/api",
+        "Avalanche": "https://api.avax.network/"
```
Comment (guanbinrui) [inline packages/web3-constants/evm/explorer.json]: It's not working. Let's leave it to empty `''`.

https://api.avax.network/?module=account&action=txlistinternal&address=0x66b57885E8E9D84742faBda0cE6E3496055b012d&startblock=0&endblock=999999999999&page=1&offset=10&sort=asc&apiKey=26GHYKIAZREN4HWJ5NCICTGKC183IGXY9A

File: packages/mask/package.json (hunk 17 above)
Comment (guanbinrui) [reply packages/mask/package.json]: Yes, will do.

**Review by amritkumarj — COMMENTED** [review]
Do I revert back to the new package.json?

File: packages/web3-constants/evm/token.json [hunk 21]
```diff
@@ -685,6 +727,7 @@
         "Arbitrum_Rinkeby": "0x0000000000000000000000000000000000000000",
         "xDai": "0x0000000000000000000000000000000000000000",
         "Celo": "0x471ece3750da237f93b8e339c536989b8978a438",
-        "Fantom": "0x0000000000000000000000000000000000000000"
+        "Fantom": "0x0000000000000000000000000000000000000000",
+        "Avalanche": "FvwEAhmxKfeiG8SnEvq42hc6whRyY3EFYAvebMqDNDGCgxN5Z"
```
Comment (guanbinrui) [inline packages/web3-constants/evm/token.json]:
```suggestion
        "Avalanche": "0x0000000000000000000000000000000000000000"
```

File: packages/web3-shared/evm/assets/chains.json [hunk 22]
```diff
@@ -2089,11 +2089,15 @@
...
         "faucets": [],
         "features": [],
-        "nativeCurrency": { "name": "Avalanche", "symbol": "AVAX", "decimals": 18 },
-        "infoURL": "https://cchain.explorer.avax.network/",
+        "nativeCurrency": {
+            "name": "Avalanche",
+            "symbol": "AVAX",
+            "decimals": 18
+        },
+        "infoURL": "https://explorer.avax.network/",
```
Comment (guanbinrui) [inline packages/web3-shared/evm/assets/chains.json]: I found snowtrace might be better. https://snowtrace.io/

File: packages/web3-constants/evm/explorer.json [hunk 23]
```diff
@@ -29,6 +30,7 @@
         "Arbitrum_Rinkeby": "",
         "xDai": "",
         "Celo": "",
-        "Fantom": ""
+        "Fantom": "",
+        "Avalanche": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/explorer.json]:
```suggestion
        "Avalanche": "JU97281F83YV3WAT6SZ4JF65JG1R5VFCF2"
```

File: packages/web3-constants/evm/explorer.json [hunk 24]
```diff
@@ -13,7 +13,8 @@
         "Arbitrum_Rinkeby": "",
         "xDai": "",
         "Celo": "",
-        "Fantom": "https://api.ftmscan.com/api"
+        "Fantom": "https://api.ftmscan.com/api",
+        "Avalanche": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/explorer.json]:
```suggestion
        "Avalanche": "https://api.snowtrace.io/api"
```

File: packages/web3-constants/evm/token-list.json [hunk 25]
```diff
@@ -27,6 +27,8 @@
         "xDai": ["https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/100/tokens.json"],
         "Celo": ["https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/42220/tokens.json"],
         "Fantom": ["https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/250/tokens.json"],
-        "Avalanche": ["https://raw.githubusercontent.com/pangolindex/tokenlists/main/ab.tokenlist.json"]
+        "Avalanche": [
+            "https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/43114/tokens.json"
```
Comment (septs) [inline packages/web3-constants/evm/token-list.json:31]: @guanbinrui switch to new domain?

**Review by guanbinrui — DISMISSED** [review]

File: packages/.eslintrc.json [hunk 26]
```diff
@@ -1,14 +1,9 @@
 {
```
Comment (septs) [inline packages/.eslintrc.json]: plz use latest version for this file.

https://github.com/DimensionDev/Maskbook/blob/develop/packages/.eslintrc.json

File: packages/mask/src/plugins/Trader/trader/useAllTradeComputed.ts [hunk 27]
```diff
@@ -187,6 +187,18 @@ export function useAllTradeComputed(
...
     const openoceanSwapEstimateGas = useDODOTradeGasLimit(openocean as TradeComputed<SwapRouteData> | null)
 
+    // traderjoe
+    const traderjoe_ = useUniswapV2Trade(
+        TradeProvider.TRADERJOE,
+        TradeStrategy.ExactIn,
+        inputAmount_,
+        '0',
+        tradeProviders.some((x) => x === TradeProvider.TRADERJOE) ? inputToken : undefined,
+        tradeProviders.some((x) => x === TradeProvider.TRADERJOE) ? outputToken : undefined,
```
Comment (septs) [inline packages/mask/src/plugins/Trader/trader/useAllTradeComputed.ts]: plz extract `tradeProviders.some((x) => x === TradeProvider.TRADERJOE)` as variable.

e.g:
```suggestion
        hasTraderJoeProvider ? inputToken : undefined,
        hasTraderJoeProvider ? outputToken : undefined,
```

Comment (amritkumarj) [reply packages/mask/src/plugins/Trader/trader/useAllTradeComputed.ts]: done

File: packages/.eslintrc.json (hunk 26 above)
Comment (amritkumarj) [reply packages/.eslintrc.json]: done

**Comment by AinuX** [comment]
@DhruvJain1122 

https://user-images.githubusercontent.com/2582974/148352998-bda17244-a586-4acd-8e0c-33f80250708a.mp4
1. Please fix the assets list and history
2. Please fix the token list

image
3. Please fix the trader plugin error 

image
4. Please fix the gas setting error

image
5. Unable to swap

https://user-images.githubusercontent.com/2582974/148354121-37d4e747-f219-4a60-bf4f-07ea305ea31f.mp4
6. Transaction status is not correct

File: packages/mask/src/components/shared/ApplicationBoard.tsx [hunk 28]
```diff
@@ -83,17 +85,18 @@ const useStyles = makeStyles()((theme) => ({
         fontSize: 15,
     },
 }))
-
-const SUPPORTED_CHAIN_ID_LIST = [
+const CHAIN_ID_LIST = [
```
Comment (septs) [inline packages/mask/src/components/shared/ApplicationBoard.tsx]: plz revert variable name

Comment (amritkumarj) [reply packages/mask/src/components/shared/ApplicationBoard.tsx]: okay

**Comment by amritkumarj** [comment]
@AinuX 
2. Token list flickering issue to this token list https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/43114/tokens.json, I have tried another token list it is working fine https://raw.githubusercontent.com/pangolindex/tokenlists/main/ab.tokenlist.json, I created a [pull request](https://github.com/DimensionDev/Mask-Token-List/pull/51) to fix the token list

3. The Trader Plugin is temporarily fixed, this issue was because of avax token address(FvwEAhmxKfeiG8SnEvq42hc6whRyY3EFYAvebMqDNDGCgxN5Z) being not comptaible, currently I have just disabled avax
4. Gas fee is also fixed

I am swapping my tokens to avax for other issues, till then please check if there is any other issue and sorry for the inconvenience

**Review by Jack-Works — DISMISSED** [review]

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/TraderView.tsx [hunk 29]
```diff
@@ -207,7 +207,8 @@ export function TraderView(props: TraderViewProps) {
         (!!trending?.coin.contract_address ||
             ['eth', 'matic', 'bnb'].includes(trending?.coin.symbol.toLowerCase() ?? '')) &&
         chainIdValid &&
-        tradeProviders.length
+        tradeProviders.length &&
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trending/TraderView.tsx]: Any problem with trading a native token on AVAX?

File: packages/mask/src/plugins/Trader/trader/useAllTradeComputed.ts [hunk 30]
```diff
@@ -187,6 +187,19 @@ export function useAllTradeComputed(
...
+    const traderjoe_ = useUniswapV2Trade(
+        TradeProvider.TRADERJOE,
+        TradeStrategy.ExactIn,
+        inputAmount_,
+        '0',
+        isTraderJoe ? inputToken : undefined,
+        isTraderJoe ? outputToken : undefined,
+    )
+    const traderjoe = useUniswapTradeComputed(traderjoe_.value, inputToken, outputToken)
+    const traderjoeEstimateGas = useUniswapTradeGasLimit(traderjoe, TradeProvider.TRADERJOE)
```
Comment (septs) [inline packages/mask/src/plugins/Trader/trader/useAllTradeComputed.ts]: extract as hook function?

Comment (amritkumarj) [reply packages/mask/src/plugins/Trader/trader/useAllTradeComputed.ts]: @septs I have updated the code, do you want this?

**Review by guanbinrui — DISMISSED** [review]

File: packages/mask/src/components/shared/ApplicationBoard.tsx [hunk 31]
```diff
@@ -83,7 +83,6 @@ const useStyles = makeStyles()((theme) => ({
         fontSize: 15,
     },
 }))
-
```
Comment (septs) [inline packages/mask/src/components/shared/ApplicationBoard.tsx]: plz revert the line.

Comment (amritkumarj) [reply packages/mask/src/components/shared/ApplicationBoard.tsx]: Done

**Review by guanbinrui — DISMISSED** [review]

**Review by Jack-Works — APPROVED** [review]

**Comment by UncleBill** [comment]
Submitted a PR to https://github.com/trustwallet/assets/pull/17549 to fix the missing token icon.

**Comment by guanbinrui** [comment]
close by #5289.

Hi @DhruvJain1122, thanks for the awesome. This PR duplicates with #5289 since the team decides to accept that request so close this one.
